arch/cortex-m/src/lib.rs

```diff
@@ -220,18 +219,25 @@ pub unsafe extern "C" fn switch_to_user_arm_v7m(
     mut user_stack: *const usize,
     process_regs: &mut [usize; 8],
 ) -> *const usize {
-    llvm_asm!(
-        "
+    asm!(
+    "
+    // Rust `asm!()` macro (as of Feb 2021) will not let us mark r6 as a
+    // clobber. However, in the process of restoring and saving the process's
+    // registers, we do in fact clobber r6. So, we work around this by doing our
+    // own manual saving of r6 using r3, and then mark r3 as clobbered.
+    mov r3, r6
```

Comment:
This makes me a little uneasy. I am not sure if it is valid to write to a reserved register in this way (r6, which LLVM uses as the base pointer on ARM). From the Rust inline asm docs:

> In some cases LLVM will allocate a "reserved register" for reg operands even though this register cannot be explicitly specified. Assembly code making use of reserved registers should be careful since reg operands may alias with those registers. Reserved registers are:
>
> - The frame pointer on all architectures.
> - r6 on ARM.

Looking into this a little further, I found [this](https://github.com/rust-lang/rust/issues/73450#issuecomment-650463347) discussion, which seems quite similar. It sounds like saving the value on the stack is the recommended approach -- I also see that approach recommended [here](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=11807#c25) in the gcc issue discussing how to mark reserved registers as clobbered.

Reply:
We can save the register to the stack, but then what? Don't we have to write it back to r6 still?

Reply:
Yes that's true, we would have to pop it back into the register. and I don't really know why that would be any different than what you suggested here. This just seems like somewhere where inline assembly can be a bit of a minefield of confusing or unexpected interactions with the compiler. I don't totally understand the warnings here or how LLVM will use r6.

Reply:
Also, we are storing r6 on the stack via r3 and the svc call, correct? r0-r3 have to be stored somewhere, right?

Reply:
yeah, I guess so. I'm a bit out of my depth here, so I'll resolve this.